apps/bf_temp.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2018 BaishanCloud. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <openssl/bio.h>
+#include "apps.h"
+
+#define BIO_TYPE_TEMP_FILE      (25|BIO_TYPE_FILTER)
+static BIO_METHOD *methods_tempfile = NULL;
+
+typedef struct {
+    char *filename;
+    char *tmp_filename;
+} TEMPFILE_CTX;
+
+static int tempfile_new(BIO *b);
+static int tempfile_free(BIO *b);
+static int tempfile_read(BIO *b, char *out, int outl);
+static int tempfile_write(BIO *b, const char *in, int inl);
+static long tempfile_ctrl(BIO *b, int cmd, long num, void *ptr);
+static int tempfile_gets(BIO *b, char *buf, int size);
+static int tempfile_puts(BIO *b, const char *str);
+static void make_temp_filename(const char *orig, char *filename);
+
+BIO_METHOD *apps_bf_tempfile(void)
+{
+    if (methods_tempfile == NULL) {
+        methods_tempfile = BIO_meth_new(BIO_TYPE_TEMP_FILE,
+                                        "Temporary file filter");
+        if (methods_tempfile == NULL
+            || !BIO_meth_set_write(methods_tempfile, tempfile_write)
+            || !BIO_meth_set_read(methods_tempfile, tempfile_read)
+            || !BIO_meth_set_puts(methods_tempfile, tempfile_puts)
+            || !BIO_meth_set_gets(methods_tempfile, tempfile_gets)
+            || !BIO_meth_set_ctrl(methods_tempfile, tempfile_ctrl)
+            || !BIO_meth_set_create(methods_tempfile, tempfile_new)
+            || !BIO_meth_set_destroy(methods_tempfile, tempfile_free))
+            return NULL;
+    }
```

Comment:
May I suggest adding `OPENSSL_atexit(apps_bf_tempfile_cleanup)` here, so callers don't need to bother with it?
(you can make `apps_bf_tempfile_cleanup` static and remove its declaration from `apps.h`)

Reply:
Awesome suggestion!